[TRTLLM-8129][feat] Allreduce tuning and benchmark script revising

hyukn opened the pull request:
Because we have encountered some perf regression due to using a one-shot kernel instead of NCCL on A100/H100, it will be beneficial if we can have a solid benchmarking of allreduce Op and analyze the data collected from it.
 
Implemented new AllreduceOp heuristic:
* Added Linear programming-based heuristic implementation.
* Added LUT-based heuristic implementation and corresponding code generation script.

AllreduceOp minor fixing:
* Fixed a minor issue in AllreduceOp, that the strategy can not be overridden when ONESHOT or TWOSHOT is set.
* Fixed a minor TWOSHOT kernel perf issue.
* Cleaned up Dispatching code in AllReduceOp.

This PR will fix the perf gaps reported in:
* https://nvbugspro.nvidia.com/bug/5517023

For Deepseek-R1 perf, it shows 3-4% perf gain in concurrencies of 256 and 512.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - All-reduce microbenchmark now uses shape-driven sweeps across versions/fusion/strategies and can export results to CSV via a new CLI flag.
  - Enhanced profiling: labeled regions, optional CUDA graphs, warmups, and median-based timing.

- Refactor
  - Strategy selection for all-reduce now strictly honors the configured strategy (no remapping).

- Tests
  - Updated benchmark validations to use direct comparisons with tolerance for improved correctness checks.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!--
Please write the PR title by following this template:

**[JIRA ticket/NVBugs ID/GitHub issue/None][type] Summary**

Valid ticket formats:
  - JIRA ticket: [TRTLLM-1234] or [FOOBAR-123] for other FOOBAR project
  - NVBugs ID: [https://nvbugs/1234567]
  - GitHub issue: [#1234]
  - No ticket: [None]

Valid types (lowercase): [fix], [feat], [doc], [infra], [chore], etc.

Examples:
  - [TRTLLM-1234][feat] Add new feature
  - [https://nvbugs/1234567][fix] Fix some bugs
  - [#1234][doc] Update documentation
  - [None][chore] Minor clean-up

Alternative (faster) way using CodeRabbit AI:

**[JIRA ticket/NVBugs ID/GitHub issue/None] @coderabbitai title**

NOTE: "@coderabbitai title" will be replaced by the title generated by CodeRabbit AI, that includes the "[type]" and title.
For more info, see /.coderabbit.yaml.

-->

## Description

<!--
Please explain the issue and the solution in short.
-->

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- The reviewers assigned automatically/manually are appropriate for the PR.


- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

hyukn commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #19343](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19343/) [ run ] triggered by Bot

hyukn commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #19554](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19554/) [ run ] triggered by Bot

hyukn committed:
[TRTLLM-8129][feat] Allreduce tuning and benchmark script revising

Modify allreduce benchmark to avoid invalid shapes, with some minor revising.

Implement new AllreduceOp heuristic:
* Add Linear programming based heuristic implementation.
* Add LUT-based heuristic implementation and corresponding code generation script.

AllreduceOp minor fixing:
* Fix a minor issue in allreduceOp that strategy can not be overridden when ONESHOT or TWOSHOT is set.
* Fix a TWOSHOT perf issue.
* Clean up Dispatching code.

Signed-off-by: Yukun He <23156053+hyukn@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #19559](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19559/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #19554](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19554/) [ run ] completed with state `ABORTED`
<a href="https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/14700/" target="_blank">LLM/main/L0_MergeRequest_PR #14700</a> <a href="https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/14700//display/redirect" target="_blank">(Blue Ocean)</a> completed with status: ABORTED

tensorrt-cicd commented:
[PR_Github #19738](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19738/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #19746](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19746/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #19738](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19738/) [ run ] completed with state `ABORTED`
<a href="https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/14858/" target="_blank">LLM/main/L0_MergeRequest_PR #14858</a> <a href="https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/14858//display/redirect" target="_blank">(Blue Ocean)</a> completed with status: ABORTED

tensorrt-cicd commented:
[PR_Github #19857](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19857/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #19857](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/19857/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #14941](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/14941/) completed with status: 'FAILURE'

hyukn commented:
/bot run --disable-fail-fast --only-multi-gpu-test

tensorrt-cicd commented:
[PR_Github #20723](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20723/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #20723](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20723/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #15658](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/15658/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #20750](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20750/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #20750](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20750/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #15681](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/15681/) completed with status: 'SUCCESS'

hyukn commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #21395](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/21395/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #21503](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/21503/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #21503](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/21503/) [ run ] completed with state `SUCCESS`
[/LLM/release-1.1/L0_MergeRequest_PR pipeline #156](https://nv/trt-llm-cicd/job/release-1.1/job/L0_MergeRequest_PR/156/) completed with status: 'SUCCESS'

litaotju commented on tests/microbenchmarks/all_reduce.py:
```diff
@@ -46,139 +134,143 @@ def allreduce_benchmark(dtype: str,
+    shape_list = []
+
+    if explore_2d:
+        num_seqs_list = [
```
A nit. Can fix later.

`num_tokens_list`?

hyukn replied:
Confirmed. Will fix it in the following PRs.

hyukn commented on tests/microbenchmarks/all_reduce.py:
```diff
@@ -27,15 +29,101 @@
+    strategy=AllReduceStrategy.NCCL,
+    fusion=AllReduceFusionOp.NONE,
+    input=None,
+    residual=None,
```
Sure. Will fix it with the following PRs.
